### PR TITLE
Fix i256 checked multiplication

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Part of #2637 

# Rationale for this change
 

The previous version was failing to correctly detect overflow.

# What changes are included in this PR?

There may be a smart way to detect signed overflow correctly, but trying to think about this just did my head in. The simple solution is to take the absolute value, perform the arithmetic and convert back. I think this is fine, and if someone has a use-case where this becomes a major performance bottleneck we can revisit.

# Are there any user-facing changes?

No

### CODE REVIEW

File: arrow-buffer/src/bigint.rs [hunk 1]
```diff
@@ -179,16 +207,29 @@ impl i256 {
...
-        let hl = self.high.checked_mul(other.low as i128)?;
-        let lh = (self.low as i128).checked_mul(other.high)?;
+        let hl = (self.high as u128).checked_mul(other.low)?;
+        let lh = self.low.checked_mul(other.high as u128)?;
 
-        Some(Self {
-            low,
-            high: (high as i128).checked_add(hl)?.checked_add(lh)?,
-        })
+        let high: i128 = high.checked_add(hl)?.checked_add(lh)?.try_into().ok()?;
```
Comment (viirya) [inline arrow-buffer/src/bigint.rs:226]: I may miss it. But don't we need to count in self.high multiplies other.high?

**Comment by tustvold** [comment]
There be more bugs :cry:

File: arrow-buffer/src/bigint.rs [hunk 2]
```diff
@@ -361,8 +411,27 @@ mod tests {
 
             let checked = il.checked_mul(ir);
             match overflow {
-                true => assert!(checked.is_none()),
-                false => assert_eq!(checked.unwrap(), actual),
+                true => assert!(
```
[the remark this answers is not in the document]
Comment (tustvold) [reply arrow-buffer/src/bigint.rs]: I added some explicit test cases

File: arrow-buffer/src/bigint.rs (hunk 1 above)
Comment (tustvold) [reply arrow-buffer/src/bigint.rs:232]: Yup, the result of the high products is necessarily overflow, and so can be ignored for the unchecked version, but must be checked here

**Comment by tustvold** [comment]
I plan to give this another once over with fresh eyes in the morning, and then get this in.

**Review by viirya — APPROVED** [review]

File: arrow-buffer/src/bigint.rs [hunk 3]
```diff
@@ -303,8 +350,106 @@ mod tests {
...
+        }
+    }
+
     #[test]
     fn test_i256() {
+        let candidates = [
+            i256::from_parts(0, 0),
+            i256::from_parts(0, 1),
+            i256::from_parts(0, -1),
+            i256::from_parts(u128::MAX, 1),
```
Comment (alamb) [inline arrow-buffer/src/bigint.rs:436]: thanks -- this is much easier to ensure coverage

**Pull request merged**